service/storage_service.cc

```diff
@@ -1088,11 +1082,24 @@ void storage_service::handle_state_normal(inet_address endpoint) {
     if (_gossiper.uses_host_id(endpoint)) {
         auto host_id = _gossiper.get_host_id(endpoint);
         auto existing = _token_metadata.get_endpoint_for_host_id(host_id);
+        auto replace_addr = db().local().get_replace_address();
         if (db().local().is_replacing() &&
-            db().local().get_replace_address() &&
-                _gossiper.get_endpoint_state_for_endpoint_ptr(db().local().get_replace_address().value())  &&
-            (host_id == _gossiper.get_host_id(db().local().get_replace_address().value()))) {
-            slogger.warn("Not updating token metadata for {} because I am replacing it", endpoint);
+            replace_addr &&
+            _gossiper.get_endpoint_state_for_endpoint_ptr(*replace_addr) &&
+            host_id == _gossiper.get_host_id(*replace_addr)) {
+            slogger.warn("Not updating token metadata for {} because I am replacing it, myip={}, node={} becomes normal status, existing node={}, node to be replaced={}",
+                    endpoint, get_broadcast_address(), endpoint, existing, *replace_addr);
+            // If the replacing node becomes NORMAL status, we can remove the
+            // node to be replaced from token_metadata. However, if the
+            // replacing node has the same ip address as the node to be
+            // replaced, we should not remove the node itself from
+            // token_metadata.
+            if (endpoint == get_broadcast_address() && *replace_addr != get_broadcast_address()) {
+               slogger.info("Remove the node to be replaced={} from token_metadata because the replacing node={} has become NORMAL status",
+                        *replace_addr, get_broadcast_address());
+                _token_metadata.remove_endpoint(*replace_addr);
```

Comment:
How do we prevent this node from learning about the node being replaced from gossip again? I.e. how do we prevent this:
- we remove it from our _token_metadata
- but immediately after, `handle_status_normal` inserts it back again

Reply:
Nodes is added to endpoints_to_remove in storage_service::handle_state_normal -> storage_service::remove_endpoint() -> gossiper::remove_endpoint() -> gossiper::quarantine_endpoint(endpoint) -> node is in _just_removed_endpoints -> In gossiper::apply_state_locally, nodes in _just_removed_endpoints will be skipped.

Reply:
Ok, but I'm wondering why the `_gossiper.replacement_quarantine(ep)` call below can be removed now.

Reply:
I guess you meant why we can remove  replacement_quarantine here. 

```
@@ -1164,10 +1171,6 @@ void storage_service::handle_state_normal(inet_address endpoint) {
 
     for (auto ep : endpoints_to_remove) {
         remove_endpoint(ep);
-        auto replace_addr = db().local().get_replace_address();
-        if (db().local().is_replacing() && replace_addr && *replace_addr == ep) {
-            _gossiper.replacement_quarantine(ep); // quarantine locally longer than normally; see CASSANDRA-8260
-        }
     }
```

Because the replacing node now removes replaced node the same way as the rest of the nodes in the cluster, all in the storage_service::handle_state_normal handler when the replacing node becomes NORMAL status.  No special treatment is needed any more.